warehouse/admin/views/projects.py

```diff
@@ -136,9 +144,108 @@ def project_detail(project, request):
         "ONE_GB": ONE_GB,
         "MAX_PROJECT_SIZE": MAX_PROJECT_SIZE,
         "UPLOAD_LIMIT_CAP": UPLOAD_LIMIT_CAP,
+        "observation_kinds": ObservationKind,
+        "observations": observations,
     }
 
 
+@view_config(
+    route_name="admin.project.observations",
+    renderer="admin/projects/observations_list.html",
+    permission="moderator",
+    request_method="GET",
+    uses_session=True,
+)
+def observations_list(project, request):
+    try:
+        page_num = int(request.params.get("page", 1))
+    except ValueError:
+        raise HTTPBadRequest("'page' must be an integer.") from None
+
+    observations_query = (
+        request.db.query(project.Observation)
+        .options(joinedload(project.Observation.observer))
+        .filter(project.Observation.related == project)
+        .order_by(project.Observation.created.desc())
+    )
+
+    # TODO: When implementing DataTables, consider removing server-side pagination.
+    #  DataTables can handle up to about 50,000 rows.
+    observations = SQLAlchemyORMPage(
+        observations_query,
+        page=page_num,
+        items_per_page=25,
+        url_maker=paginate_url_factory(request),
+    )
+
+    return {"observations": observations, "project": project}
+
+
+@view_config(
+    route_name="admin.project.add_observation",
+    permission="moderator",
+    request_method="POST",
+    uses_session=True,
+    require_methods=False,
+)
+def add_observation(project, request):
+    kind = request.POST.get("kind")
+    if not kind:
+        request.session.flash("Provide a kind", queue="error")
+        raise HTTPSeeOther(
+            request.route_path(
+                "admin.project.detail", project_name=project.normalized_name
+            )
+        )
+
+    kind_map = {kind.value[0]: kind for kind in ObservationKind}
+
+    try:
+        kind = kind_map[kind]
+    except KeyError as e:
+        request.session.flash("Invalid kind", queue="error")
+        raise HTTPSeeOther(
+            request.route_path(
+                "admin.project.detail", project_name=project.normalized_name
+            )
+        ) from e
+
+    summary = request.POST.get("summary")
+    if not summary:
+        request.session.flash("Provide a summary", queue="error")
+        raise HTTPSeeOther(
+            request.route_path(
+                "admin.project.detail", project_name=project.normalized_name
+            )
+        )
+
+    # We allow an empty payload from Admin.
+    payload = {}
+
+    # TODO: This probably shouldn't be here.
+    #  It has to do with the association creation pattern
+    #  I haven't figured out yet.
+    from warehouse.observations.models import Observer
+
+    if not request.user.observer:
+        request.user.observer = Observer()
+
+    project.record_observation(
+        request=request,
+        kind=kind,
+        observer=request.user.observer,
+        summary=summary,
+        payload=payload,
+    )
```

Comment:
Wondering if we're going down the same path as https://github.com/pypi/warehouse/issues/8090 here by storing these on the project (or release) and not on individual files. I see how it would be tedious to report N observations for N files on a release, or M releases in a project, but this sort of prevents us from being able to make observations about individual artifacts.

Reply:
Excellent observation (pun totally intended).

The overall intent is that the mixin be flexible enough to record Observations at any level - Project, Release, File, User, Organization, etc.

I've only added the mixin so far to the Project and Releases, but only added the Admin UI for Projects.

Any mixed-in model can then call `<model>.record_observation(...)` to add a record.

Reply:
@di I've added an example of adding observations to Releases via Admin as well to hopefully make it clearer.